### PR TITLE
Fixes #32265 - Support unrecognized services in ping

### PR DESCRIPTION
**Pull request by ofedoren**
@ekohl, probably you want to review/test this as well.

This should add a warning message in case if a service failed and is not listed in the output.

ScreenShot-1602617557330

### CODE REVIEW

File: lib/hammer_cli_foreman/command_extensions/ping.rb [hunk 1]
```diff
@@ -10,9 +10,28 @@ class Ping < HammerCLI::CommandExtensions
...
+        warn [_('%{count} more service(s) failed, but not shown:') % { count: failed.size },
+              failed.keys.join(', '),
+              ''].join("\n")
+      end
+
       def self.failed?(services)
-        services.each_value.any? { |s| s['status'] == _('FAIL') } ||
-          services['foreman']['database']['active'] == 'FAIL'
+        services['foreman']['database']['active'] == 'FAIL' ||
+          services.each_value.any? { |s| s['status'] == _('FAIL') }
```
Comment (ekohl) [inline lib/hammer_cli_foreman/command_extensions/ping.rb:34]: Is there anything changing here? It looks like you're just swapping the order.

File: lib/hammer_cli_foreman/command_extensions/ping.rb [hunk 2]
```diff
@@ -10,9 +10,28 @@ class Ping < HammerCLI::CommandExtensions
...
+                        .each_with_object([]) { |(_, d), s| s << d['services'] }
+                        .reduce({}, :merge)
+                        .select do |name, data|
+          begin
+            output_definition.find_field(name)
+            false
+          rescue ArgumentError
+            data['status'] == _('FAIL')
+          end
+        end
```
Comment (ekohl) [inline lib/hammer_cli_foreman/command_extensions/ping.rb:24]: This logic is quite hard to follow. Why does it modify data in place? Shouldn't it be a more pure function that returns a list of unrecognized services?

File: lib/hammer_cli_foreman/ping.rb [hunk 3]
```diff
@@ -19,9 +19,12 @@ def execute
         response = send_request
         print_data(response)
 
-        return 1 if HammerCLIForeman::CommandExtensions::Ping.failed?(response)
+        return HammerCLI::EX_OK unless HammerCLIForeman::CommandExtensions::Ping.failed?(response)
 
-        HammerCLI::EX_OK
```
Comment (ekohl) [inline lib/hammer_cli_foreman/ping.rb:24]: I like the exit if it's abnormal and closing with `EX_OK`. That implies `EX_OK` is the expected execution. Your version implies exiting with 1 is normal.

File: lib/hammer_cli_foreman/command_extensions/ping.rb (hunk 1 above)
Comment (ofedoren) [reply lib/hammer_cli_foreman/command_extensions/ping.rb:34]: yeap, exactly. That is just to "short-circuit" checks.

File: lib/hammer_cli_foreman/command_extensions/ping.rb (hunk 2 above)
Comment (ofedoren) [reply lib/hammer_cli_foreman/command_extensions/ping.rb:24]: This method doesn't modify anything. If there is a failed service that has no field (won't be shown) it simply puts a warning. Otherwise does nothing additional.

File: lib/hammer_cli_foreman/ping.rb (hunk 3 above)
Comment (ofedoren) [reply lib/hammer_cli_foreman/ping.rb:24]: Hmm... I thought that exit with `EX_OK` `unless` something failed should be done right away and don't check for which services will be shown and which won't.

I can refactor that, but I'm not sure if it will be more readable though..

Comment (ekohl) [reply lib/hammer_cli_foreman/ping.rb:24]: When I read a function (and this was also what previously happened) is that I'd expect all steps in between would error out early.

So roughly:
```ruby
return 1 if failure_condition_1
return 2 if failure_condition_2

# Everything went OK

0
```

This makes it easier to extend with more failure conditions. In this case, I'd suggest:
```ruby
if HammerCLIForeman::CommandExtensions::Ping.failed?(response)
  HammerCLIForeman::CommandExtensions::Ping.check_for_unrecognized(
     response, output_definition
  )
  return 1
end

HammerCLI::EX_OK
```

IMHO the flow is easier to follow: we see a failure, do some additional processing and return early. Does that make sense or am I explaining it poorly?

Comment (ofedoren) [reply lib/hammer_cli_foreman/ping.rb:24]: @ekohl, I just was going to update the PR :D

Yes, that makes sense.

**Comment by ofedoren** [comment]
@ekohl, updated with your suggestion.

**Review by ekohl — APPROVED** [review]
Thanks!

**Comment by ekohl** [comment]
@ofedoren I have merge permissions because I'm an admin, not because I'm a regular maintainer. Do you expect another hammer maintainer to review or shall I hit the big green button?

**Comment by ofedoren** [comment]
@ekohl, well let's try to ask @yifatmakias :) Yifat, if you are interested or have time, could you please take a look?

**Comment by yifatmakias** [comment]
Looks good to me. Thanks @ofedoren :)
